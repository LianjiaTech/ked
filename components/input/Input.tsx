import * as React from 'react';
import * as PropTypes from 'prop-types';
import { polyfill } from 'react-lifecycles-compat';
import classNames from 'classnames';
import omit from 'omit.js';
import Group from './Group';
import Search from './Search';
import TextArea from './TextArea';
import Password from './Password';
import { Omit, tuple } from '../_util/type';
import ClearableLabeledInput, { hasPrefixSuffix } from './ClearableLabeledInput';
import { ConfigConsumer, ConfigConsumerProps } from '../config-provider';
import warning from '../_util/warning';

export const InputSizes = tuple('small', 'default', 'large');

export interface InputProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size' | 'prefix'> {
  prefixCls?: string;
  size?: typeof InputSizes[number];
  onPressEnter?: React.KeyboardEventHandler<HTMLInputElement>;
  addonBefore?: React.ReactNode;
  addonAfter?: React.ReactNode;
  prefix?: React.ReactNode;
  suffix?: React.ReactNode;
  allowClear?: boolean;
}

export function fixControlledValue<T>(value: T) {
  if (typeof value === 'undefined' || value === null) {
    return '';
  }
  return value;
}

export function resolveOnChange(
  target: HTMLInputElement | HTMLTextAreaElement,
  e:
    | React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
    | React.MouseEvent<HTMLElement, MouseEvent>,
  onChange?: (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => void,
) {
  if (onChange) {
    let event = e;
    if (e.type === 'click') {
      // click clear icon
      event = Object.create(e);
      event.target = target;
      event.currentTarget = target;
      const originalInputValue = target.value;
      // change target ref value cause e.target.value should be '' when clear input
      target.value = '';
      onChange(event as React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>);
      // reset target ref value
      target.value = originalInputValue;
      return;
    }
    onChange(event as React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>);
  }
}

export function getInputClassName(
  prefixCls: string,
  size?: typeof InputSizes[number],
  disabled?: boolean,
) {
  return classNames(prefixCls, {
    [`${prefixCls}-sm`]: size === 'small',
    [`${prefixCls}-lg`]: size === 'large',
    [`${prefixCls}-disabled`]: disabled,
  });
}

export interface InputState {
  value: any;
}

class Input extends React.Component<InputProps, InputState> {
  static Group: typeof Group;

  static Search: typeof Search;

  static TextArea: typeof TextArea;

  static Password: typeof Password;

  static defaultProps = {
    type: 'text',
  };

  static propTypes = {
    type: PropTypes.string,
    id: PropTypes.string,
    size: PropTypes.oneOf(InputSizes),
    maxLength: PropTypes.number,
    disabled: PropTypes.bool,
    value: PropTypes.any,
    defaultValue: PropTypes.any,
    className: PropTypes.string,
    addonBefore: PropTypes.node,
    addonAfter: PropTypes.node,
    prefixCls: PropTypes.string,
    onPressEnter: PropTypes.func,
    onKeyDown: PropTypes.func,
    onKeyUp: PropTypes.func,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    prefix: PropTypes.node,
    suffix: PropTypes.node,
    allowClear: PropTypes.bool,
  };

  input: HTMLInputElement;

  clearableInput: ClearableLabeledInput;

  removePasswordTimeout: number;

  constructor(props: InputProps) {
    super(props);
    const value = typeof props.value === 'undefined' ? props.defaultValue : props.value;
    this.state = {
      value,
    };
  }

  static getDerivedStateFromProps(nextProps: InputProps) {
    if ('value' in nextProps) {
      return {
        value: nextProps.value,
      };
    }
    return null;
  }

  componentDidMount() {
    this.clearPasswordValueAttribute();
  }

  // Since polyfill `getSnapshotBeforeUpdate` need work with `componentDidUpdate`.
  // We keep an empty function here.
  componentDidUpdate() {}

  getSnapshotBeforeUpdate(prevProps: InputProps) {
    if (hasPrefixSuffix(prevProps) !== hasPrefixSuffix(this.props)) {
      warning(
        this.input !== document.activeElement,
        'Input',
        `When Input is focused, dynamic add or remove prefix / suffix will make it lose focus caused by dom structure change. Read more: https://ant.design/components/input/#FAQ`,
      );
    }
    return null;
  }

  componentWillUnmount() {
    if (this.removePasswordTimeout) {
      clearTimeout(this.removePasswordTimeout);
    }
  }

  focus() {
    this.input.focus();
  }

  blur() {
    this.input.blur();
  }

  select() {
    this.input.select();
  }

  saveClearableInput = (input: ClearableLabeledInput) => {
    this.clearableInput = input;
  };

  saveInput = (input: HTMLInputElement) => {
    this.input = input;
  };

  setValue(value: string, callback?: () => void) {
    if (!('value' in this.props)) {
      this.setState({ value }, callback);
    }
  }

  handleReset = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
    this.setValue('', () => {
      this.focus();
    });
    resolveOnChange(this.input, e, this.props.onChange);
  };

  renderInput = (prefixCls: string) => {
    const { className, addonBefore, addonAfter, size, disabled } = this.props;
    // Fix https://fb.me/react-unknown-prop
    const otherProps = omit(this.props, [
      'prefixCls',
      'onPressEnter',
      'addonBefore',
      'addonAfter',
      'prefix',
      'suffix',
      'allowClear',
      // Input elements must be either controlled or uncontrolled,
      // specify either the value prop, or the defaultValue prop, but not both.
      'defaultValue',
      'size',
      'inputType',
    ]);
    return (
      <input
        {...otherProps}
        onChange={this.handleChange}
        onKeyDown={this.handleKeyDown}
        className={classNames(getInputClassName(prefixCls, size, disabled), {
          [className!]: className && !addonBefore && !addonAfter,
        })}
        ref={this.saveInput}
      />
    );
  };

  clearPasswordValueAttribute = () => {
    // https://github.com/ant-design/ant-design/issues/20541
    this.removePasswordTimeout = window.setTimeout(() => {
      if (
        this.input &&
        this.input.getAttribute('type') === 'password' &&
        this.input.hasAttribute('value')
      ) {
        this.input.removeAttribute('value');
      }
    });
  };

  handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.setValue(e.target.value, this.clearPasswordValueAttribute);
    resolveOnChange(this.input, e, this.props.onChange);
  };

  handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const { onPressEnter, onKeyDown } = this.props;
    if (e.keyCode === 13 && onPressEnter) {
      onPressEnter(e);
    }
    if (onKeyDown) {
      onKeyDown(e);
    }
  };

  renderComponent = ({ getPrefixCls }: ConfigConsumerProps) => {
    const { value } = this.state;
    const { prefixCls: customizePrefixCls } = this.props;
    const prefixCls = getPrefixCls('input', customizePrefixCls);
    return (
      <ClearableLabeledInput
        {...this.props}
        prefixCls={prefixCls}
        inputType="input"
        value={fixControlledValue(value)}
        element={this.renderInput(prefixCls)}
        handleReset={this.handleReset}
        ref={this.saveClearableInput}
      />
    );
  };

  render() {
    return <ConfigConsumer>{this.renderComponent}</ConfigConsumer>;
  }
}

polyfill(Input);

export default Input;
