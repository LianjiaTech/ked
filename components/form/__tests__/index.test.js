/* eslint-disable react/prefer-stateless-function */
import React from 'react';
import { mount, render } from 'enzyme';
import Form from '..';
import Input from '../../input';
import mountTest from '../../../../tests/shared/mountTest';
import './type.test';

describe('Form', () => {
  mountTest(Form);
  mountTest(Form.Item);

  it('hideRequiredMark', () => {
    const wrapper = mount(<Form hideRequiredMark />);
    expect(wrapper.find('form').hasClass('ant-form-hide-required-mark')).toBe(true);
  });

  it('Form.Item should support data-*、aria-* and custom attribute', () => {
    const wrapper = render(
      <Form>
        <Form.Item data-text="123" aria-hidden="true" cccc="bbbb">
          text
        </Form.Item>
      </Form>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('wrappedComponentRef', () => {
    it('warns on functional component', () => {
      if (process.env.REACT === '15') {
        return;
      }
      const spy = jest.spyOn(console, 'error').mockImplementation(() => {});
      const TestForm = () => <Form />;
      const Wrapped = Form.create()(TestForm);
      mount(<Wrapped wrappedComponentRef={() => {}} />);
      expect(spy).toHaveBeenCalled();
      spy.mockReset();
      spy.mockRestore();
    });

    it('get component ref', () => {
      class TestForm extends React.Component {
        // eslint-disable-line
        render() {
          return <Form />;
        }
      }
      const Wrapped = Form.create()(TestForm);
      let form;
      mount(
        <Wrapped
          wrappedComponentRef={node => {
            form = node;
          }}
        />,
      );
      expect(form).toBeInstanceOf(TestForm);
    });
  });

  describe('FormItem', () => {
    it('FormItem: generate snapshot when validates fields', async done => {
      let wrapper;
      const TestForm = props => (
        <Form>
          <Form.Item>
            {props.form.getFieldDecorator('test', {
              rules: [
                {
                  validator: (rule, value, callback) => {
                    setTimeout(() => {
                      callback();
                      expect(wrapper.render()).toMatchSnapshot(); // after validate
                      done();
                    }, 100);
                  },
                },
              ],
            })(<Input allowClear />)}
          </Form.Item>
        </Form>
      );
      const Wrapped = Form.create()(TestForm);
      wrapper = mount(<Wrapped />);
      expect(wrapper.render()).toMatchSnapshot(); // before validate
      wrapper.find('.ant-input').simulate('change', { target: { value: 'test' } });
      await new Promise(resolve => setTimeout(resolve, 0));
      expect(wrapper.render()).toMatchSnapshot(); // validating
    });
  });
});
