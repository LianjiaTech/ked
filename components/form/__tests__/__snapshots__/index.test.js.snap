// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Form.Item should support data-*„ÄÅaria-* and custom attribute 1`] = `
<form
  class="ant-form ant-form-horizontal"
>
  <div
    aria-hidden="true"
    cccc="bbbb"
    class="ant-row ant-form-item"
    data-text="123"
  >
    <div
      class="ant-col ant-form-item-control-wrapper"
    >
      <div
        class="ant-form-item-control"
      >
        <span
          class="ant-form-item-children"
        >
          text
        </span>
      </div>
    </div>
  </div>
</form>
`;

exports[`Form FormItem FormItem: generate snapshot when validates fields 1`] = `
<form
  class="ant-form ant-form-horizontal"
>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-control-wrapper"
    >
      <div
        class="ant-form-item-control"
      >
        <span
          class="ant-form-item-children"
        >
          <span
            class="ant-input-affix-wrapper"
          >
            <input
              class="ant-input"
              data-__field="[object Object]"
              data-__meta="[object Object]"
              id="test"
              type="text"
              value=""
            />
            <span
              class="ant-input-suffix"
            />
          </span>
        </span>
      </div>
    </div>
  </div>
</form>
`;

exports[`Form FormItem FormItem: generate snapshot when validates fields 2`] = `
<form
  class="ant-form ant-form-horizontal"
>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-control-wrapper"
    >
      <div
        class="ant-form-item-control is-validating"
      >
        <span
          class="ant-form-item-children"
        >
          <span
            class="ant-input-affix-wrapper"
          >
            <input
              class="ant-input"
              data-__field="[object Object]"
              data-__meta="[object Object]"
              id="test"
              type="text"
              value="test"
            />
            <span
              class="ant-input-suffix"
            >
              <i
                aria-label="icon: close-circle"
                class="anticon anticon-close-circle ant-input-clear-icon"
                role="button"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="close-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 0 1-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                  />
                </svg>
              </i>
            </span>
          </span>
        </span>
      </div>
    </div>
  </div>
</form>
`;

exports[`Form FormItem FormItem: generate snapshot when validates fields 3`] = `
<form
  class="ant-form ant-form-horizontal"
>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-control-wrapper"
    >
      <div
        class="ant-form-item-control has-success"
      >
        <span
          class="ant-form-item-children"
        >
          <span
            class="ant-input-affix-wrapper"
          >
            <input
              class="ant-input"
              data-__field="[object Object]"
              data-__meta="[object Object]"
              id="test"
              type="text"
              value="test"
            />
            <span
              class="ant-input-suffix"
            >
              <i
                aria-label="icon: close-circle"
                class="anticon anticon-close-circle ant-input-clear-icon"
                role="button"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="close-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 0 1-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                  />
                </svg>
              </i>
            </span>
          </span>
        </span>
      </div>
    </div>
  </div>
</form>
`;
